tpkg/cutest/cutest_event.c

```diff
@@ -0,0 +1,364 @@
+{
+	pid_t pid;
+	int wstatus = 0;
+	struct testvars *vars = (struct testvars *)arg;
+	struct proc *proc;
+
+	assert(sig == SIGCHLD);
+	assert(event & EV_SIGNAL);
+	assert(arg != NULL);
+
+	do {
+		pid = waitpid(-1, &wstatus, WNOHANG);
+		proc = NULL;
+		if(vars->agent.pid == pid) {
+			proc = &vars->agent;
+		} else if(pid > 0) {
+			for(int i = 0; i < NUM_CHILDREN; i++) {
+				if(vars->children[i].pid == pid) {
+					proc = &vars->children[i];
+				}
+			}
+		}
+		if(proc != NULL) {
+			vars->sigchld_count++;
+			proc->wstatus = wstatus;
+			proc->events |= EV_SIGNAL;
+		}
+	} while((pid == -1 && errno == EINTR) || (pid > 0));
+
+	if(vars->sigchld_count == (NUM_CHILDREN + 1)) {
+		event_base_loopexit(vars->event_base, NULL);
+	}
+}
+
+static void
+fork_children(struct testvars *vars)
+{
+	for(int i = 0; i < NUM_CHILDREN; i++) {
+		pid_t pid;
+		int ret, fdin[2], fdout[2];
+		ret = pipe(fdin);
+		assert(ret == 0);
+		ret = pipe(fdout);
+		assert(ret == 0);
+		if((pid = fork()) == 0) {
+			close(fdin[1]);
+			close(fdout[0]);
+			child(fdin[0], fdout[1]);
+			/* never reached */
+		}
+		close(fdin[0]);
+		close(fdout[1]);
+		vars->children[i].pid = pid;
+		vars->children[i].fdin = fdin[1];
+		vars->children[i].fdout = fdout[0];
+		event_set(
+			&vars->children[i].event,
+			 vars->children[i].fdin,
+			 EV_READ | EV_CLOSED,
+			&child_callback,
+			&vars->children[i]);
+		event_base_set(vars->event_base, &vars->children[i].event);
+		event_add(&vars->children[i].event, NULL);
+	}
+}
+
+static void
+fork_agent(struct testvars *vars, void(*func)(struct proc *))
+{
+	pid_t pid;
+	int ret, fdin[2], fdout[2];
+	struct timeval timeout = { 1, 0 };
+	ret = pipe(fdin);
+	assert(ret == 0);
+	ret = pipe(fdout);
+	assert(ret == 0);
+	if((pid = fork()) == 0) {
+		close(fdin[1]);
+		close(fdout[0]);
+		agent(fdin[0], fdout[1], vars, func);
+		/* never reached */
+	}
+	close(fdin[0]);
+	close(fdout[1]);
+	vars->agent.pid = pid;
+	vars->agent.fdin = fdin[1];
+	vars->agent.fdout = fdout[0];
+	event_set(
+		&vars->agent.event,
+		 vars->agent.fdin,
+		 EV_WRITE | EV_CLOSED,
+		&agent_callback,
+		 vars);
+	event_base_set(vars->event_base, &vars->agent.event);
+	event_add(&vars->agent.event, &timeout);
+}
+
+static void
+event_setup(struct testvars *vars)
+{
+	int ret;
+
+	memset(vars, 0, sizeof(*vars));
+	vars->event_base = nsd_child_event_base();
+	assert(vars->event_base != NULL);
+	event_set(&vars->event_sigchld, SIGCHLD, EV_SIGNAL | EV_PERSIST, sigchld_callback, vars);
+	ret = event_base_set(vars->event_base, &vars->event_sigchld);
+	assert(ret == 0);
+	ret = event_add(&vars->event_sigchld, NULL);
+	assert(ret == 0);
+
+	for(int i = 0; i < NUM_CHILDREN; i++) {
+		vars->children[i].pid = -1;
+		vars->children[i].fdin = -1;
+		vars->children[i].fdout = -1;
+	}
+}
+
+static void
+event_teardown(struct testvars *vars)
+{
+	if(vars->agent.fdin != -1 &&
+	   event_initialized(&vars->agent.event) &&
+	   event_get_fd(&vars->agent.event) == vars->agent.fdin)
```

Comment:
`event_initialized` and `event_get_fd` are (likely) not supported for other libevent-API that we target.  Eg. minievent, libev, older libevent(likely), and other wrappers. Like EV_CLOSED (just use EV_READ instead, I think, for select and poll), use another variable to indicate initialization (yes this is also a source of bugs, as the initialization gets wrong, you see these kinds of variables in the NSD source, event_added. event_get_fd, I think, by fetching it from the event structure directly, `vars->agent.event.ev_fd`

Reply:
`event_initialized` is supported by libevent 1.4-2.1 and libev. minievent doesn't. I'll implemented that one for minievent. Seems useful to have it there too(?) and read the file descriptor directly from the struct.

Reply:
On second thought. Initializing the event struct with all zeroes and just checking for equality between `ev_fd` and `fdin` seems like a better idea.